Update regex literal lexing and emission

hamishknight opened the pull request:
Based on top of https://github.com/apple/swift/pull/40531, only the last commit is relevant.

---

Update the regex lexing implementation to defer to the regex library, which will pass back the pointer from which to resume lexing, and update the emission to call the new `Regex(_regexString:version:)` overload, that will accept the regex string with delimiters.

Because this uses the library's lexing implementation, the delimiters are now `'/.../'` and `'|...|'` instead of plain `'...'`.

rxwei committed:
Integrate experimental string processing modules and enable end-to-end regex.

- Checkout apple/swift-experimental-string-processing using a tag.
- Build `_MatchingEngine` as part of libswift (`ExperimentalRegex`) using sources from the package.
- Parse regex literals using the parser from `_MatchingEngine`.
- Build both `_MatchingEngine` and `_StringProcessing` as part of core libs using sources from the package.
- Use `Regex<DynamicCaptures>` as the default regex type until we finalize apple/swift-experimental-string-processing#68.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@rintaro Mind taking a look at the lexer/parser changes?

rxwei reviewed (approved):
LGTM

rintaro commented on lib/Parse/Lexer.cpp:
```diff
@@ -1951,36 +1958,31 @@ const char *Lexer::findEndOfCurlyQuoteStringLiteral(const char *Body,
+    diagnose(TokStart, diag::regex_literal_parsing_error, ErrStr);
+    formToken(tok::unknown, TokStart);
+    return true;
+  }
```
Could you add an assertion checking `CurPtr[-1]` is a valid delimiter character?

rintaro commented on lib/Parse/Lexer.cpp:
```diff
@@ -1951,36 +1958,31 @@ const char *Lexer::findEndOfCurlyQuoteStringLiteral(const char *Body,
-    return formToken(tok::unknown, TokStart);
+  // Ask libswift to try and lex a regex literal, and bail if it can't.
+  const char *ErrStr = nullptr;
+  auto *EndPtr = regexLiteralLexingFn(TokStart, BufferEnd, &ErrStr);
```
Mind adding a comment about the semantics of `nullptr` return. What if the end delimiter is missing, etc.

rintaro commented on include/swift/Parse/ExperimentalRegexBridging.h:
```diff
@@ -5,17 +5,20 @@
-
-void Parser_registerParseRegexStrawperson(ParseRegexStrawperson fn);
-bool Parser_hasParseRegexStrawperson();
+typedef const char *(* RegexLiteralLexingFn)(/*CurPtr*/ const char *,
+                                             /*BufferEnd*/ const char *,
+                                             /*ErrorOut*/ const char **);
+void Parser_registerRegexLiteralLexingFn(RegexLiteralLexingFn fn);
+
+typedef void(* RegexLiteralParsingFn)(/*InputString*/ const char *,
+                                      /*ErrorOut*/ const char **,
+                                      /*VersionOut*/ unsigned *,
+                                      /*CaptureStructureOut*/ char *,
+                                      /*CaptureStructureSize*/ unsigned);
+void Parser_registerRegexLiteralParsingFn(RegexLiteralParsingFn fn);
```
Mind adding doc-comments?

hamishknight replied:
@rintaro How does this look?

hamishknight commented:
@swift-ci please test

rintaro commented on lib/Parse/Lexer.cpp:
```diff
@@ -1951,36 +1958,35 @@ const char *Lexer::findEndOfCurlyQuoteStringLiteral(const char *Body,
+  assert(EndPtr > TokStart && EndPtr <= BufferEnd);
+  CurPtr = EndPtr;
+
+  if (ErrStr) {
+    diagnose(TokStart, diag::regex_literal_parsing_error, ErrStr);
+    formToken(tok::unknown, TokStart);
```
There might be some cases where we emit an error, but still want to form a regex literal token.

String literal has a `FIXME` where even if the end delimiter is missing, we should form a string literal and type check it, so that we can get additional semantic diagnostics, and more importantly, we can get code completion even inside malformed string interpolations. For example, consider `"Name: \(user.#HERE# "`, in this case, end delimiter is missing because it's consumed as the string interpolation, because `)` is missing.  But we still want code completion for `user`.

How about changing `regexLiteralLexingFn` like:

```cpp
const char *Ptr = TokStart;
const char *ErrStr = nullptr;

// Success: status = SUCCESS, advances `Ptr`, ErrStr == nullptr..
// Error, but parsable: status = SUCCESS, advances `Ptr` and populate `ErrStr`. 
// Error, non parsable: status = FAILED, advances `Ptr` and populate `ErrStr`. 
// Total failure: status = FAILED, Ptr == TokStart, ErrStr == nullptr.

bool status = regexLiteralLexingFn(&Ptr, BufferEnd, &ErrStr)
if (ErrStr) {
  diagnose(TokStart, diag::regex_literal_parsing_error, ErrStr);
}
if (Ptr == TokStart) {
  return false;
}
assert(Ptr > TokStart);
formToken(status ? tok::unknown : tok::regex_literal, TokStart);
return false
```

hamishknight replied:
@rintaro How does this look?

rxwei commented:
I updated my PR to fix some build script issues so you might need to rebase this. Hopefully both PRs can be merged today.

milseman commented:
@swift-ci please test macOS platform

hamishknight committed:
Update regex literal lexing and emission

Update the lexing implementation to defer to the
regex library, which will pass back the pointer
from to resume lexing, and update the emission to
call the new `Regex(_regexString:version:)`
overload, that will accept the regex string with
delimiters.

Because this uses the library's lexing
implementation, the delimiters are now `'/.../'`
and `'|...|'` instead of plain `'...'`.

hamishknight commented:
@swift-ci please test

milseman commented:
@swift-ci please test macOS platform

rintaro reviewed (approved):
Lexer/Parser changes LGTM. Thank you!

aschwaighofer commented:
Looks like this or the commit this is based on broke the incremental bot.
